controller/k8s/injector.go

```diff
@@ -0,0 +1,363 @@
+	batchV1 "k8s.io/api/batch/v1"
+	"k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+/* The v1.PodSpec struct contains a field annotation that causes the
+ * InitContainers field to be omitted when serializing the struct as json.
+ * Since we wish for this field to be included, we have to define our own
+ * enhancedPodSpec struct with a different annotation on this field.  We then
+ * must define our own structs to use this struct, and so on.
+ */
+type enhancedPodSpec struct {
+	*v1.PodSpec
+	InitContainers []v1.Container `json:"initContainers"`
+}
+
+type enhancedPodTemplateSpec struct {
+	*v1.PodTemplateSpec
+	Spec enhancedPodSpec `json:"spec,omitempty"`
+}
+
+type enhancedDeploymentSpec struct {
+	*v1beta1.DeploymentSpec
+	Template enhancedPodTemplateSpec `json:"template,omitempty"`
+}
+
+type enhancedDeployment struct {
+	*v1beta1.Deployment
+	Spec enhancedDeploymentSpec `json:"spec,omitempty"`
+}
+
+type enhancedReplicationControllerSpec struct {
+	*v1.ReplicationControllerSpec
+	Template enhancedPodTemplateSpec `json:"template,omitempty"`
+}
+
+type enhancedReplicationController struct {
+	*v1.ReplicationController
+	Spec enhancedReplicationControllerSpec `json:"spec,omitempty"`
+}
+
+type enhancedReplicaSetSpec struct {
+	*v1beta1.ReplicaSetSpec
+	Template enhancedPodTemplateSpec `json:"template,omitempty"`
+}
+
+type enhancedReplicaSet struct {
+	*v1beta1.ReplicaSet
+	Spec enhancedReplicaSetSpec `json:"spec,omitempty"`
+}
+
+type enhancedJobSpec struct {
+	*batchV1.JobSpec
+	Template enhancedPodTemplateSpec `json:"template,omitempty"`
+}
+
+type enhancedJob struct {
+	*batchV1.Job
+	Spec enhancedJobSpec `json:"spec,omitempty"`
+}
+
+type enhancedDaemonSetSpec struct {
+	*v1beta1.DaemonSetSpec
+	Template enhancedPodTemplateSpec `json:"template,omitempty"`
+}
+
+type enhancedDaemonSet struct {
+	*v1beta1.DaemonSet
+	Spec enhancedDaemonSetSpec `json:"spec,omitempty"`
+}
+
+type PodConfig struct {
+	InitImage             string
+	ProxyImage            string
+	ProxyUID              int64
+	InboundPort           uint
+	OutboundPort          uint
+	IgnoreInboundPorts    []uint
+	IgnoreOutboundPorts   []uint
+	ProxyControlPort      uint
+	ProxyAPIPort          uint
+	ProxyLogLevel         string
+	ConduitVersion        string
+	ImagePullPolicy       string
+	ControlPlaneNamespace string
+}
+
+var DefaultProxyConfig = PodConfig{
+	InitImage:             "gcr.io/runconduit/proxy-init",
+	ProxyImage:            "gcr.io/runconduit/proxy",
+	ProxyUID:              2102,
+	InboundPort:           4143,
+	OutboundPort:          4140,
+	IgnoreInboundPorts:    nil,
+	IgnoreOutboundPorts:   nil,
+	ProxyControlPort:      4190,
+	ProxyAPIPort:          8086,
+	ConduitVersion:        version.Version,
+	ImagePullPolicy:       "IfNotPresent",
+	ProxyLogLevel:         "warn,conduit_proxy=info",
+	ControlPlaneNamespace: "conduit",
```

Comment:
since most (all?) of these default values exist somewhere else in the code, can you create constants and share the values? does not have to be in this PR, you can just leave a comment and/or open a follow up ticket.